### ISSUE TITLE
Integrating 3rd-party validation libraries: Open up the design of ValidationVisitor

### ISSUE THREAD

**Issue by JeremySkinner** [aspnet/Mvc#7014]
Back in 2015, I had a discussion with @rynowak was very helpful in getting some changes into the MVC validation pipeline that would make it easier for 3rd party validation libraries to integrate with MVC's validation infrastructure (#1260), but unfortunately there are still several blocks that make this difficult. 

The main issue is that the MVC validation pipeline does not run model-level validators if any property-level validators have failed for the current model object. This behaviour is not customizable/overridable as the ValidationVisitor class is made almost entirely of nonvirtual, private methods.  

For example, the [FluentValidation](https://github.com/JeremySkinner/FluentValidation) library executes all rules in one go after model binding has occurred (ie it validates already set property values, rather than one property at a time like DataAnnotations). Unfortunately this does not work well if you mix DataAnnotations and FluentValidation in the same project. If a DataAnnotations attribute validator fails, then FluentValidation will never be executed as the property-level DataAnnotations rule prevents model-validators rules from running.

**Current workarounds:**
1) Bypass the ModelValidatorProvider Api entirely and instead implement a custom IObjectModelValidator that does the work. Not ideal as you lose many of the benefits of the ModelValidatorProvider api (eg the ability to mix multiple validation libraries)
2) Fork the code for the ValidationVisitor and maintain separately. Not ideal for obvious reasons

**Proposal**: Modify [ValidationVisitor](https://github.com/aspnet/Mvc/blob/236ef5d1d1d12a2503c6da0900b3c40ca5470985/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs) to either:
1) Add a property to control whether property-level errors should prevent model-level validators for running
2) Open up the class to make the methods protected virtual instead of private (and add protected property getters for the private fields) so 3rd party validation libraries can add this logic in themselves if option 1) is not something that you feel should be in the core framework. 

For option 1, the VisitComplexType method would be modified as follows:

```csharp
private bool VisitComplexType(IValidationStrategy defaultStrategy)
{
    var isValid = true;

    if (_model != null && _metadata.ValidateChildren)
    {
        var strategy = _strategy ?? defaultStrategy;
        isValid = VisitChildren(strategy);
    }
    else if (_model != null)
    {
        // Suppress validation for the entries matching this prefix. This will temporarily set
        // the current node to 'skipped' but we're going to visit it right away, so subsequent
        // code will set it to 'valid' or 'invalid'
        SuppressValidation(_key);
    }

    
    // Original check
    //if (isValid && !_modelState.HasReachedMaxErrors)
    //{
    //    isValid &= ValidateNode();
    //}
    
    // Proposed new check:
    if ((isValid || ValidateComplexTypesIfChildValidationFails) && !_modelState.HasReachedMaxErrors)
    {
        isValid &= ValidateNode();
    }

    return isValid;
}

public bool ValidateComplexTypesIfChildValidationFails {get;set;}
```

If going with option 2), then all the private methods including VisitComplexType should be made protected virtual, and protected properties with getters would be needed to expose the private fields to subclasses.

I would be happy to submit a pull request for either approach depending on what's preferred.

**Comment by JeremySkinner** [comment]
Would be great to get a response to this. /cc @rynowak @Eilon

**Comment by rynowak** [comment]
Sorry for the delay, I think both of your proposals are reasonable. If you'd prefer to do just proposal one (`ValidateComplexTypesIfChildValidationFails`) that's fine too.

Feel free to send a PR, and if you don't have a chance, I will get to this before we ship 2.1

**Comment by JeremySkinner** [comment]
Thanks, I've created a pull request that implements both proposals and adds tests. Let me know if you'd like me to make any changes.

**Comment by JeremySkinner** [comment]
I've moved the properties below the ctor as requested. Something went wrong with the PR when merging latest changes from dev so I've recreated the PR (#7075)

**Comment by rynowak** [comment]
Fixed by: 9c59e63 a06d0d0 0ccfe4c 5a6bc3a
